simdistserve/base/worker.py

```diff
@@ -182,28 +197,74 @@ def forward_decode(self, items: Union['Request', Iterable['Request']], to_schedu
         _decode_len = min(remaining_tok_in_batch, len(self.decode_queue))
-        decode_reqs = []
-        for i in range(_decode_len):
-            req = self.decode_queue[0]
-            if (req.current_context_len + 1) > decode_max_tokens:
-                break
-            decode_max_tokens -= (req.current_context_len + 1)
-            decode_reqs.append(self.decode_queue.popleft())
+        decode_reqs: 'List[Request]' = []
+    
+        # if memory is not enough, only schedule the requests that have been decoded before
+        # because their kv-cache is already alloced, new requests' kv-cache cost too much free memory    
+        decode_all_kinds_requests = self.mem_is_enough()
+
+        # request is given up if 
+        # 1. request needs kv-cache migrate but memory is less than the mem_slot_lower_bound
+        # 2. available memory is not enough
+        requests_give_up = deque()
+        
+        if self.mem_is_enough() or decode_all_kinds_requests:
+            available_slots = self.free_mem_slots_num // 2 # avoid free_mem_slots_num is used in one batch
+        else:
+            available_slots = self.free_mem_slots_num // 128 # batch size control
+        
+        for _ in range(_decode_len):
+            left_req = self.decode_queue[0]
+            self.decode_queue.popleft()
+            if not left_req.kvcache_is_transferred: # requests just finished prefill
+                if not decode_all_kinds_requests:
+                    requests_give_up.append(left_req) # needs kv-cache migrate, give up
+                    continue
+                elif available_slots > left_req.current_context_len : # memory is enough, migrate kv-cache
+                    available_slots -= left_req.current_context_len
+                    decode_reqs.append(left_req)
+                else: # memory is not enough, give up the request
+                    requests_give_up.append(left_req)
+            else: # common requests
+                if available_slots <= 1:
+                    requests_give_up.append(left_req)
+                    continue
+                decode_reqs.append(left_req)
+                available_slots -= 1
+
+
+            
+        # put the requests kicked back to the queue from front
+        while len(requests_give_up) > 0:
+            self.decode_queue.appendleft(requests_give_up.pop())
+        assert len(self.decode_queue) + len(decode_reqs) == _decode_len
+            
+        # kv-cache transfer
+        migrate_time = 0
+        for r in list(decode_reqs):
+            if not r.kvcache_is_transferred: 
+                if r.prefill_worker.wid != self.wid: # if the kv-cache is not in the worker, then migrate
+                    migrate_time += self.per_token_kvcache_transfertime * r.current_context_len
+                self.migrate_alloc_kvcache([r,])
+                r.prefill_worker.wakeup()
+            
+        yield self.env.timeout(migrate_time)
+        self.decode_alloc_kvcache(decode_reqs)
```

Comment:
[Major] OK for now - but this also means kv cache migration is a blocking operation. In DistServe, KV cache migration is non-blocking. 

Better impl is to have another loop `migrate_request()` that:
```bash
while True:
  if self.migrate_queue.is_empty()
    yield self.wait_for_event() # I forgot - search simpy's Event-based wake up
  # now migration queue has stuff
  migration_time = ...
  yield self.env.timeout(migration_time)
  # the rest of the logic to:
  # 1. free kv from prefill and add kv to decode
  # 2. add request to the decode worker waiting queue?
  self.decode_worker.wakeup() # or something similar that let the decode know the waiting queue has stuff added in

```